### DISCUSSION TITLE
Column view like in MC?

### DISCUSSION THREAD

**Discussion by aloneguid** [tui-cs/Terminal.Gui#3006] [Q&A] [answered]
Midnight Commander has 2 panes, i.e. one of the panes looks like this:

![image](https://github.com/gui-cs/Terminal.Gui/assets/3155189/f4922fdd-c6d7-43d7-9a77-6367832b7896)

Can you point me in the right direction which control I can use to implement this? ListView seems like it's for rendering a one-line item. And GridView requires DataGrid which I'd rather not use. Thanks!

**Comment by migueldeicaza** [comment]
You can just customize what gets drawn by the ListView so whatever you desire

**Reply by aloneguid** [reply]
Thank you! I'm manually printing vertical lines as a part of item rendering, looks fine so far: 
![image](https://github.com/gui-cs/Terminal.Gui/assets/3155189/bde32f1f-34bc-436c-b106-9cadacdd1111)

**Reply by tznind** [reply, marked as answer]
In v2_develop I removed the coupling of `TableView` from `DataTable`.  You can define your own columns with `EnumerableTableSource<>`.  You can use the alpha nuget package or link the source repo directly for the latest changes.

```csharp
	Application.Init ();
	var w = new Window ();

	var currentDir = new DirectoryInfo (Environment.CurrentDirectory);
	
	var fsi = currentDir.GetFileSystemInfos ();
	var source = new EnumerableTableSource<FileSystemInfo> (fsi,
		new Dictionary<string, Func<FileSystemInfo, object>> {
			{ "Name",(f)=>f.Name},
			{ "Size",(f)=>f is FileInfo fi ? fi.Length : ""},
			{ "Mod",(f)=>f.LastWriteTime},
		});

	var tv = new TableView (source) {
		Width = Dim.Fill (),
		Height = Dim.Fill ()
	};

	w.Add (tv);
	Application.Run (w);
	Application.Shutdown ();
```

![image](https://github.com/gui-cs/Terminal.Gui/assets/31306100/ac327504-b2e4-4975-a2c2-57a37c8c437a)

You can even create a tree table (see `OpenTreeExample` in [TableEditor example](https://github.com/gui-cs/Terminal.Gui/blob/0d1bbdd0f9293b8c7c13acc86f623af44f3023fa/UICatalog/Scenarios/TableEditor.cs#L657C16-L657C32)):

![tree-table](https://github.com/gui-cs/Terminal.Gui/assets/31306100/b4e4d562-23d6-4cfb-99db-aaa27513e2ba)

**Reply by aloneguid** [reply]
Thank you, i managed to do wht I wanted. This library is magic.

**Reply by tznind** [reply]
Glad you like it! :)
